tests/lang_callable/varres/ttemplate_var_lent_res.nim

```diff
@@ -0,0 +1,47 @@
+discard """
+  description: '''
+    Ensure that templates with `var` and `lent` return types behave the same
+    as normal procedures would
+  '''
+"""
+
+# XXX: they currently don't. Refer to the assertion with the
+#      "now works" string
+
+block scalar_type:
+  # test with a scalar type
+  template borrowVar(x: untyped): var int = x
+  template borrowLent(x: untyped): lent int = x
+
+  var v = 0
+  borrowVar(v) = 1 # test that an assignment works
+  doAssert borrowVar(v) == 1 # test that a read works
+  doAssert compiles((borrowLent(v) = 2)), "now works"
+  doAssert borrowLent(v) == 1
+
+block tuple_type:
+  # test with a tuple type
+  type Tuple = tuple[x: int]
+
+  template borrowVar(x: untyped): var Tuple = x
+  template borrowLent(x: untyped): lent Tuple = x
+
+  # note: `v` being an unnamed tuple is deliberate, as it is
+  # intended to ensure that the implicit conversion in the template
+  # doesn't cause problems
+  var v: (int,) = (0,)
+  # test full assignment and read access:
+  borrowVar(v) = (x: 1)
+  doAssert borrowVar(v) == (x: 1)
+  # test with field write and read:
+  borrowVar(v).x = 2
```

Comment:
mostly a note for myself, but having this work by clobbering the type is a bit mind bending and possibly dangerous.

Reply:
What do you mean by "clobbering the type"?

Reply:
If i'm following the expansion correctly, `borrowVar(v)` is just `v`. `v`'s type is `var (,int)`, but by way of the template it's `var tuple`, which due to assignment(?) is treated as `var (, x: int)`, allowing for `borrowVar(v).x`, even though `v` originally didn't have an `x` field.

Reply:
Ah, that you mean. Unnamed tuples are currently implicitly convertible to named ones that have the same structure, but the type of `v` stays the same. The expanded code essentially looks like this:
```nim
Tuple(v).x = 2
```

Relying on this behaviour is not necessary for the test, so I've now changed `v` to be of type `Tuple`.